File: src/main/java/com/github/spartusch/rateretriever/rate/v1/provider/RateProvider.java
package com.github.spartusch.rateretriever.rate.v1.provider;

import reactor.core.publisher.Mono;

import java.math.BigDecimal;

public interface RateProvider {

    Mono<BigDecimal> getCurrentRate(final String symbol, final String currencyCode);

    Mono<Boolean> isCurrencyCodeSupported(final String currencyCode);

}

File: src/main/kotlin/com/github/spartusch/rateretriever/rate/v1/provider/RateProvider.kt
package com.github.spartusch.rateretriever.rate.v1.provider

import com.github.spartusch.rateretriever.rate.v1.model.ProviderId
import com.github.spartusch.rateretriever.rate.v1.model.TickerSymbol
import io.micrometer.core.instrument.MeterRegistry
import javax.money.CurrencyUnit
import javax.money.MonetaryAmount

interface RateProvider {
    fun getProviderId(): ProviderId

    fun getCurrentRate(
        symbol: TickerSymbol,
        currency: CurrencyUnit
    ): MonetaryAmount?

    fun isCurrencySupported(
        currency: CurrencyUnit
    ): Boolean
}

abstract class AbstractTimedRateProvider(private val meterRegistry: MeterRegistry) : RateProvider {
    protected val requestTimer by lazy {
        meterRegistry.timer("provider.requests", "provider.name", getProviderId().toString())
    }
}
